### ISSUE TITLE
Exposing more data from the api-calls

### ISSUE THREAD

**Issue by tobbensoft** [zabuldon/teslajsonpy#52]
During debugging to fix #51 and seeing the available data from the api..   

Is it a big thing to expose this in HA without any "processing" just so we could start using it to display information?

That could probably bring more interest and contributors to the project.

**Comment by alandtse** [comment]
Exposing anything in HA would be difficult because of the review process in pushing code. If you describe what you want to specifically have access to in HA that would be easier. There are a bunch of sites that allow you to browser the known [Tesla API](https://www.teslaapi.io/). The question is to consider how to represent data in HA.

**Comment by tobbensoft** [comment]
Listing a few from the top of my mind that could be useful

Last successful api-call (timestamp) in combination with state: online/asleep
charge_limit_soc_std: 
charge_state: 
car_version:

Managing online/asleep would be a good thing (that I do not know enough about yet) 
The services @llluis was mentioning he implemented in #24 is probably in the right direction.

**Comment by alandtse** [comment]
Thanks. Please think about how they would be displayed in HA. We can easily expose the data you want through the API as it's all collected but the question will be how to properly display it in HA. I have a [charging rate sensor PR](https://github.com/home-assistant/home-assistant/pull/30286) to try to get more info for charging so I'm not sure if that's what you needed.

**Comment by tobbensoft** [comment]
@alandtse your PR looks great..  

awake/sleep could be represented by a binary_sensor.tesla_awake ?  (I could not find any of the device classes the seems to fit awake/sleep)

Is there any guidelines when sensors don't fit devices classes?  Like car_version: for instance?

**Comment by magnussand** [comment]
I just started out using HA and are currently setting up the development environment for it.

I'm also interested in this as well as have the ability to read remaining charge time, charge current and turn on or off individual seat heaters.

I guess things like on/off of each seat heater should be switches, but everything that is read only seems logical to have registered as sensors, for instance, current location, speed, cabin temperature etc.

**Comment by tobbensoft** [comment]
@alandtse your latest PR works fine in the pre-release!

**Comment by alandtse** [comment]
@magnussand We did expose a bunch of the charger related stuff. We don't have the seat heaters yet, but do check the component as some of the other temperature stuff and gps stuff is already exposed.

@alandtse Glad to hear. Waiting for it to go live so I can benefit from the work on my production system. ;)

**Comment by alandtse** [comment]
@tobbensoft Online binary sensor can be easily added by replacing two files in teslajsonpy.
See [https://github.com/zabuldon/teslajsonpy/compare/master...alandtse:online_sensor](url). HA 0.104.0 will automatically populate it once teslajsonpy is updated.  Also, no idea why you have 2 volts when disconnected, that's what Tesla is reporting 🤷‍♂.

**Comment by tobbensoft** [comment]
Upgraded to HA 0.104.0 and replaced the files..  loads and populates as expected.

Is binary_sensor correct for the online status?

To my knowledge it has at least 3 states. online/asleep/offline

**Comment by llluis** [comment]
Thank guys for the effort.
To allow a state machine through automations as proposed by @tobbensoft, the online binary sensor shall have a separate behaviour from the rest of the API: it shall poll /api/1/vehicles/{id} at short intervals (like, 60s, configurable maybe). This call never wakes the car and would allow the automations to change poll interval of the full data dynamically.

Problem is: I never succeeded in changing the poll interval in HA during runtime. As far as I saw, it's always 30s but with throttle constraint.

Or we can implement this state machine in teslajsonpy. Would be a lot easier and more manageable, avoiding HA approval cycles as mentionned by @alandtse.

**Comment by tobbensoft** [comment]
@llluis by the looks of the logs in "teslafi" it seems to be 60s for the short interval polling.

And I assume you mean 300s :)

**Comment by alandtse** [comment]
> Is binary_sensor correct for the online status?
> Is there a default naming for the status connected/disconnected
> 
> To my knowledge it has at least 3 states. online/asleep/offline

Yes, a connector switch has those default naming. Templates will respond with `On` and `Off` though. Given we aren't exposing an explicit wake command, what value are you getting if you know it's asleep vs offline? For example, what would you do differently? Please remember commands automatically will wake up the car while updates will not.

> To allow a state machine through automations as proposed by @tobbensoft, the online binary sensor shall have a separate behaviour from the rest of the API: it shall poll /api/1/vehicles/{id} at short intervals (like, 60s, configurable maybe). This call never wakes the car and would allow the automations to change poll interval of the full data dynamically.

This is the behavior more or less (but using 300 instead of 60). The tesljsonpy controller updates the online state every `scan_interval` seconds. After the online check, it then checks the Update switch to decide if it'll directly ping the car for any cars that are determined to be online.

So to changes in HA, if you have a `scan_interval` of 300 (which is currently the hardcoded floor), you could decide to have HA skip updates for a vehicle some period after you detect the parking brake is engaged. For example, assume you'll be in a location for an hour after driving.  You then can enable the update switch when you want to start getting updates again (but it has to be engaged for at least `scan_interval` so it's picked up during one of the controller update cycles).

> Problem is: I never succeeded in changing the poll interval in HA during runtime. As far as I saw, it's always 30s but with throttle constraint.

Ignore the HA logs for updates. That 30s polling is coming from HA and will get the cached value. The controller operates at `scan_interval` to update the cache per my explanation above.

Now there could be a question about letting the `scan_interval` go even lower by removing the 300 floor. We have always been conservative to avoid getting people locked out of their car or having Tesla decide to lock down their API.

**Comment by tobbensoft** [comment]
@alandtse will try to catch the offline mode and match it with the sensor state. (only seen it a few times)

Is it possible to get a timestamp when last successful API-call was made? (I like to know that everything is ok and polling when car is asleep) 
I can't think of how to use the information present in HA for that right now.

**Comment by alandtse** [comment]
> Is it possible to get a timestamp when last successful API-call was made? (I like to know that everything is ok and polling when car is asleep)
No. That data isn't saved anywhere. You'll only see it in teslajsonpy logs. I'm not aware of any HA property or attribute where we'd save that info as the philosophy is the api is supposed to handle things like that. You can assume it's polling per my description above and you can check the logs.

**Comment by llluis** [comment]
> This is the behavior more or less (but using 300 instead of 60). The tesljsonpy controller updates the online state every `scan_interval` seconds. After the online check, it then checks the Update switch to decide if it'll directly ping the car for any cars that are determined to be online.

I thought that the online status were polled at the same time as everything else. That means that if the car is sleeping, independent of the online switch, it will detect it became online? Because what we need is to detect the car sucessfully slept to reduce polling time and detect a drive or charge right away. I need to make a few tests.

> So to changes in HA, if you have a `scan_interval` of 300 (which is currently the hardcoded floor), you could decide to have HA skip updates for a vehicle some period after you detect the parking brake is engaged. For example, assume you'll be in a location for an hour after driving. You then can enable the update switch when you want to start getting updates again (but it has to be engaged for at least `scan_interval` so it's picked up during one of the controller update cycles).

Online status could always be 60s. Only the car requests would be afftected by scan_interval. But I believe we can live with as is.
My idea was to remove the floor and set it to 60s. Then, the only situation it would increase (15 min) is when you want to let the car sleep (parking break engaged for 15min). At 15min interval, car will sleep inbetween. After that you can return to 60s and will catch the wakeup.

> Ignore the HA logs for updates. That 30s polling is coming from HA and will get the cached value. The controller operates at `scan_interval` to update the cache per my explanation above.

Yeah, I forgot the throttle is done at the controller and not in the update method in the HA integration. Good, then we can expose the "scan_interval" to dynamically change it through a service call.

> Now there could be a question about letting the `scan_interval` go even lower by removing the 300 floor. We have always been conservative to avoid getting people locked out of their car or having Tesla decide to lock down their API.

My TeslaFi polls it every 60s, and so far so good, no problem of being kicked out. But it's dynamic. If it detects a situation of "can sleep", it changes to 15 min. Once the car slept, it goes back to 60s to catch the wake up. If driving or charing, it kleeps the 60s.

**Comment by llluis** [comment]
> @llluis by the looks of the logs in "teslafi" it seems to be 60s for the /api/1/vehicles/{id} polling.

@tobbensoft, Yes, Teslafi is always checking online status every 60s.

> And I assume you mean 300s :)
HA calls the controller method every 30s. But the controller throttles this request to scan_interval, which is 300s. So the car is only called every 300s.

**Comment by alandtse** [comment]
> I thought that the online status were polled at the same time as everything else. That means that if the car is sleeping, independent of the online switch, it will detect it became online?

Yes, it will detect if online if Tesla says it's online. The Update switch does not do anything to the online check.

> Good, then we can expose the "scan_interval" to dynamically change it through a service call.
> 
> >

No need to expose a new service, just change the config flow option. However, no idea if the config_flow options are changeable programmatically and that would be a HA feature request if it isn't. That said, you can just turn off the Update switch for periods you want the car to be sleeping. You don't have to change the `scan_interval`.

**Comment by tobbensoft** [comment]
@alandtse I think I got the update/sleeping working as it should by using the update switch and automations. Will test for a little longer.  (Maybe write a little FAQ for the docs?)

2 things are not perfect yet though :)

- Detection of the online state is limited to the 300s interval.  That could almost be a whole drive to the kindergarten for me. 
As discussed above..  why don't we make the online check 60s and keep the other check configurable at 300s or above? (Teslafi is using this polling rate and works fine) (It seems that the polling limit is per token and not per car)

- With a restart of HA and the car is in sleep state, a lot of data is "unavailable" and screws up the log/graphs..  This is not the integrations fault but if we wakeup the car after a reboot, the data would be correct... Are there any other solutions?  (Template seems wrong here I think)
What are your thoughts here?

**Comment by alandtse** [comment]
Not opposed to allowing someone to set the online check to 60 but should keep the default. Please note there are people with multiple cars so we can't set the default too low.

There is a forced wake up on restart so assuming the cars respond, it will populate. Of course we can't force cars to respond so templates will need to wait for data to populate. Some templates use a uptime sensor and require at least 5 minutes of HA uptime before they activate.

**Comment by tobbensoft** [comment]
- Sounds good, how is the online check configurable?

- I have done multiple restarts when car is asleep and never seen that it wakes up or the car is sent any wakeup command.  Any thoughts how to check whats wrong?

**Comment by alandtse** [comment]
> * Sounds good, how is the online check configurable?

It's the scan_interval. We need to up the floor, but that can be 60. Your template can then manage how often any vehicles are polled. If we really want to expose to scan_intervals, see if the HA constants has more than one else we'll have to create a new variable which will need to be approved by HA.

> * I have done multiple restarts when car is asleep and never seen that it wakes up or the car is sent any wakeup command.  Any thoughts how to check whats wrong?

`teslajsonpy` logs will show you the API calls its making. On restart it will update all data with a forced wakeup  and if the car is asleep it will retry the wakeup/data pull 5 times with an exponential delay until it gives up.   Look for the initial online check `https://owner-api.teslamotors.com/api/1/vehicles` and then the json response to see whether the car is online. Next you'll see the individual car pull. `https://owner-api.teslamotors.com/api/1/vehicles/XXX/data`.   The response will tell you if it responded right away or had to try a wakeup routine.

**Comment by tobbensoft** [comment]
Can I hack some file to change ```scan_interval```  right now?

> There is a forced wake up on restart so assuming the cars respond, it will populate. Of course we can't force cars to respond so templates will need to wait for data to populate. Some templates use a uptime sensor and require at least 5 minutes of HA uptime before they activate.

What I see in the logs is that the ```https://owner-api.teslamotors.com/api/1/vehicles``` call is made, but I can't see any wakeup command initiated? (If i toggle something, everything works as expected)

So the ```https://owner-api.teslamotors.com/api/1/vehicles/XXX/data``` is never made.

I can try this again tonight to make sure I'm not missing anything.

**Comment by alandtse** [comment]
Probably makes sense to message me on discord instead of trying to chat in github. `alandtse#5464`

> Can I hack some file to change `scan_interval` right now?
> 
> >

1. Change the `clamp` value here and restart HA.
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/tesla/config_flow.py#L104

2. And then update the options in the integration to whatever you want.

> So the `https://owner-api.teslamotors.com/api/1/vehicles/XXX/data` is never made.

Ok, there's a bug in the code.
https://github.com/zabuldon/teslajsonpy/blob/master/teslajsonpy/controller.py#L481

```python
            async with self.__lock[vin]:
                if (
                    online
                    and (  # pylint: disable=too-many-boolean-expressions
                        self.__update.get(vin)
                    )
```
should be:
```python
            async with self.__lock[vin]:
                if (
                    (online or wake_if_asleep)
                    and (  # pylint: disable=too-many-boolean-expressions
                        self.__update.get(vin)
                    )```
Try that and confirm the behavior for me.

**Comment by tobbensoft** [comment]
Tried to restart when car is asleep.

It seems that it detects that it needs to wake up the car but does not issue the api-call for it..  

````
2020-01-23 19:06:23 DEBUG (MainThread) [teslajsonpy.connection] post: https://owner-api.teslamotors.com/oauth/token
2020-01-23 19:06:26 DEBUG (MainThread) [teslajsonpy.connection] {"access_token": "349864f40916d6929d3faf6bc4fe1e7aa31d0a655a7ec7105803fc0789b8b7d9", "token_type": "bearer", "expires_in": 3888000, "refresh_token": "201449cd738af4a1cbbccd0089bdc0868a93dcd7f286fdb3c79253b2aac41338", "created_at": 1579802786}
2020-01-23 19:06:26 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/vehicles
2020-01-23 19:06:27 DEBUG (MainThread) [teslajsonpy.connection] {"response": [{"id": 5025388345911957, "vehicle_id": <vehicle_id>, "vin": "<vin>", "display_name": "Old Grey", "option_codes": "AD15,MDL3,PBSB,RENA,BT37,ID3W,RF3G,S3PB,DRLH,DV2W,W39B,APF0,COUS,BC3B,CH07,PC30,FC3P,FG31,GLFR,HL31,HM31,IL31,LTPB,MR31,FM3B,RS3H,SA3P,STCP,SC04,SU3C,T3CA,TW00,TM00,UT3P,WR00,AU3P,APH3,AF00,ZCST,MI00,CDM0", "color": null, "tokens": ["e29b97885689b811", "48fa49cb10b518f8"], "state": "asleep", "in_service": false, "id_s": "5025388345911957", "calendar_enabled": true, "api_version": 7, "backseat_token": null, "backseat_token_updated_at": null}], "count": 1}
2020-01-23 19:06:27 DEBUG (MainThread) [teslajsonpy.controller] wake_up needed for get -> None 
Info: args:(<teslajsonpy.controller.Controller object at 0x7fa1c2db03d0>, 5025388345911957, 'data', True), kwargs:{}, car_id:5025388345911957, car_online:{'<vin>': False}
2020-01-23 19:06:27 DEBUG (MainThread) [teslajsonpy.controller] Vehicle is awake, trying function get
2020-01-23 19:06:27 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/vehicles/5025388345911957/data
2020-01-23 19:06:27 DEBUG (MainThread) [teslajsonpy.connection] {"response": null, "error": "vehicle unavailable: {:error=>\"vehicle unavailable:\"}", "error_description": ""}
2020-01-23 19:06:27 DEBUG (MainThread) [teslajsonpy.controller] get(5025388345911957): Retry Attempt(0): False
2020-01-23 19:06:31 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/vehicles/5025388345911957/data
2020-01-23 19:06:31 DEBUG (MainThread) [teslajsonpy.connection] {"response": null, "error": "vehicle unavailable: {:error=>\"vehicle unavailable:\"}", "error_description": ""}
2020-01-23 19:06:31 DEBUG (MainThread) [teslajsonpy.controller] get(5025388345911957): Retry Attempt(1): False
2020-01-23 19:06:39 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/vehicles/5025388345911957/data
2020-01-23 19:06:39 DEBUG (MainThread) [teslajsonpy.connection] {"response": null, "error": "vehicle unavailable: {:error=>\"vehicle unavailable:\"}", "error_description": ""}
2020-01-23 19:06:39 DEBUG (MainThread) [teslajsonpy.controller] get(5025388345911957): Retry Attempt(2): False
2020-01-23 19:06:55 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/vehicles/5025388345911957/data
2020-01-23 19:06:56 DEBUG (MainThread) [teslajsonpy.connection] {"response": null, "error": "vehicle unavailable: {:error=>\"vehicle unavailable:\"}", "error_description": ""}
2020-01-23 19:06:56 DEBUG (MainThread) [teslajsonpy.controller] get(5025388345911957): Retry Attempt(3): False
2020-01-23 19:07:28 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/vehicles/5025388345911957/data
2020-01-23 19:07:29 DEBUG (MainThread) [teslajsonpy.connection] {"response": null, "error": "vehicle unavailable: {:error=>\"vehicle unavailable:\"}", "error_description": ""}
2020-01-23 19:07:29 DEBUG (MainThread) [teslajsonpy.controller] get(5025388345911957): Retry Attempt(4): False
2020-01-23 19:11:33 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/vehicles
2020-01-23 19:11:33 DEBUG (MainThread) [teslajsonpy.connection] {"response": [{"id": 5025388345911957, "vehicle_id": <vehicle_id>, "vin": "<vin>", "display_name": "Old Grey", "option_codes": "AD15,MDL3,PBSB,RENA,BT37,ID3W,RF3G,S3PB,DRLH,DV2W,W39B,APF0,COUS,BC3B,CH07,PC30,FC3P,FG31,GLFR,HL31,HM31,IL31,LTPB,MR31,FM3B,RS3H,SA3P,STCP,SC04,SU3C,T3CA,TW00,TM00,UT3P,WR00,AU3P,APH3,AF00,ZCST,MI00,CDM0", "color": null, "tokens": ["e29b97885689b811", "48fa49cb10b518f8"], "state": "asleep", "in_service": false, "id_s": "5025388345911957", "calendar_enabled": true, "api_version": 7, "backseat_token": null, "backseat_token_updated_at": null}], "count": 1} ```

Sent a msg on discord..

**Comment by alandtse** [comment]
Added some additional fixes in https://github.com/zabuldon/teslajsonpy/compare/master...alandtse:online_sensor. Will probably push that as a new version over the weekend. If you could let me know if it breaks anything, that would be great.

**Comment by tobbensoft** [comment]
Installed... looks good so far...

**Comment by llluis** [comment]
Everything all good so far. I'm experimenting a few other changes and your branch is great. 

I've changed the online check to 60s and independent of the main polling:
> https://github.com/llluis/teslajsonpy/blob/aeada5cca1c322d029c2a034eacecd8ca5fc36a9/teslajsonpy/controller.py#L472

Seems to be working pretty good. I put some additional logging to verify.
Creating a few automations now to enable and disable the update switch.

Few changes so far: https://github.com/alandtse/teslajsonpy/compare/online_sensor...llluis:master

**Comment by alandtse** [comment]
Please test this version. https://github.com/zabuldon/teslajsonpy/pull/57. Websockets should connect and automatically update various drive states every few seconds if driving is ever detected (during regular scan_interval check). Unfortunately, HA will still only update the device tracker at 30 second intervals as the websocket update is transparent and I didn't figure out how to increase poll rate for the device tracker only.

**Comment by tobbensoft** [comment]
Running #57 , will only do one drive today..  But will try to keep an eye on what's happening..

**Comment by llluis** [comment]
I merged #57 and added control to switch from 60 to 600 polling along with automations to control it. I believe it will be the way to do it.
Problem with turning off update switch is that until the car sleeps, you will loose everything inbetween. With a 10min polling, you give the car 10 min slices to try to sleep. 

Also, I loved the attributes in the controls. I added a few more, awesome!!! Gives all the info without creating additional entities. Even if needed, templates will solve it.
